src/ui/components/Widgets/Chart/hooks/useIntersectionObserver.ts

```diff
@@ -0,0 +1,129 @@
+import React from 'react';
+
+import {throttle} from 'lodash';
+
+type IntersectionCallback = (state: boolean) => void;
+
+const LOADING_VISIBLE_OFFSET = 300;
+
+class Observer {
+    callbacksMap: Map<HTMLDivElement, IntersectionCallback> = new Map();
+    changesQueue: Map<HTMLDivElement, boolean> | null = null;
+    intersectionObserver: IntersectionObserver | null = null;
+
+    intersectionFlushQueue = throttle(() => {
+        if (this.changesQueue === null) {
+            return;
+        }
+
+        const {callbacksMap, changesQueue} = this;
+        const changes = changesQueue;
+        this.changesQueue = null;
+
+        changes.forEach((isVisible, el) => {
+            callbacksMap.get(el)?.(isVisible);
+        });
+        changes.clear();
+    }, 100);
+
+    intersectionHandler = (entries: IntersectionObserverEntry[]) => {
+        this.changesQueue = this.changesQueue || new Map();
+
+        const hasChanges = entries.reduce((memo, e) => {
+            const target = e.target as HTMLDivElement;
+            const state = e.intersectionRatio > 0;
```

Comment:
bool variables better name close to `isVisible` or `hasIntersection` or something like this. Name `state` can confuse